File: 102/binary-tree-level-order-traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func levelOrder(root *TreeNode) (res [][]int) {
	if nil == root {
		return res
	}
	nodes := []*TreeNode{root}
	var vals []int
	for len(nodes) > 0 {
		nodes, vals = nextLevelNodesAndVals(nodes)
		res = append(res, vals)
	}
	return res
}

func nextLevelNodesAndVals(nodes []*TreeNode) (nextNodes []*TreeNode, vals []int) {
	for _, node := range nodes {
		if nil == node {
			continue
		}
		vals = append(vals, node.Val)
		if nil != node.Left {
			nextNodes = append(nextNodes, node.Left)
		}
		if nil != node.Right {
			nextNodes = append(nextNodes, node.Right)
		}
	}
	return nextNodes, vals
}

File: 102/BinaryTreeLevelOrderTraversal.scala
object BinaryTreeLevelOrderTraversal {
  def levelOrder(root: TreeNode): List[List[Int]] = {
    @annotation.tailrec
    def levelOrder(currLevel: List[TreeNode], res: List[List[Int]]): List[List[Int]] =
      if (currLevel.isEmpty)
        res
      else {
        var nextLevel = currLevel.foldRight(List.empty[TreeNode]) ( (node, ls) => {
          if (node.left != null && node.right != null)
            node.left :: node.right :: ls
          else if (node.left != null)
            node.left :: ls
          else if (node.right != null)
            node.right :: ls
          else
            ls
        })
        levelOrder(nextLevel, currLevel.map(_.value) :: res)
      }

    if (root != null)
      levelOrder(List(root), List()).reverse
    else
      Nil
  }

  class TreeNode(var _value: Int = 0) {
    var value: Int = _value
    var left: TreeNode = null
    var right: TreeNode = null
  }
}
